README
iPlant-RFactory

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc iPlant::RFactory

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=iPlant-RFactory

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/iPlant-RFactory

    CPAN Ratings
        http://cpanratings.perl.org/d/iPlant-RFactory

    Search CPAN
        http://search.cpan.org/dist/iPlant-RFactory/


LICENSE AND COPYRIGHT

Copyright (C) 2013 Naim Matasci

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for iPlant-RFactory

0.01    Date/time
        First version, released on an unsuspecting world.



t/data/BayesA.R
BayesA <-function(y,Z,startpi,startdf,startscale,truepi, truedef, truescale,numiter,skip,burnIn,Seed=1000,Save.at="results")  
#  startpi is defined by the proportion of markers that ARE associate with genes
{   # START OF FUNCTION

	  set.seed(Seed)  

	  library(msm)

	  if(numiter <=burnIn) stop ("Total Number of MCMC iterations should exceed number of burnin iterations")

	  #filename=paste("LDsimu_",Seed,"_",2,sep="")
	  #data(filename)

	  #Save.at=paste("Change to your directory",sep=" ")
  
	  # Parameters
	  nSNP=dim(Z)[2]  #number of SNP
	  nanim=dim(Z)[1]; #number of animals

	  # Vectors for saved samples of hyperparameters
	  if (!truedef)    {defsave   =  array(0,numiter/skip) }
	  if (!truescale)  {scalesave =  array(0,numiter/skip) }
	  if (!truepi)     {pisave    =  array(0,numiter/skip) } 
	  varesave = array(0,numiter/skip)

	  # To monitor MH acceptance ratios on degrees of freedom parameter
	  alphadef_save = array(0,(numiter-burnIn))

	  # input data
	  nrecords = length(y)
	  X = array(1,nrecords)
	  dimX = 1
	  W = as.matrix(cbind(X,Z))
	  rm(Z)
	  Wstacked=as.vector(W);

	  # specify prior degrees of belief on genetic variance
	  # specify  prior scale on genetic variance
	  # inital values   

	  def=startdf			 # Starting Bayes A df 
	  scalea  = startscale    	 # Starting scale parameter
	  pi = startpi                   # Starting pi parameter   (probability of being a gene!)                                                          
	  cdef       = 0.5		 # cdef is scale parameter for proposal density on df
	  alphacheck = 0
	  alphaQTLvar= 0			 # to monitor BayesB QTLvar rejection rates 
	  alphatot1= 0
	  WWdiag     = apply(W,2,crossprod) # diagonals are important

	  # prior for residual variance
	  nu_e     = -1          # i.e. corresponding to Gelman's prior.
	  scale_e  = 0

	  # used to monitor Metropolis Hastings acceptance ratio for df
	  dimW     = dim(W)[2]  # number of columns = # of location parameters

	  # initialize a few arrays
	  theta      = array(0,nSNP+1)
	  SNPeff     = array(0,nSNP)
	  varq       = array(scalea,nSNP)  # initialize each SNP variance to overall scale parameter
	  SNPvar     = array(0,nSNP)       # used to take running sum of SNP specific variances
	  SNPvarlast = array(0,nSNP)
	  postprob   = array(0,nSNP)       # indicator variable for inclusion of SNP in model
	                                   # later used to determine posterior probability...unique to BayesB
	  alphametrovar = array(0,nSNP)    # used to monitor MH acceptance rates for included genes.  
  
	  # This is for computing mean and variance of each SNP effect.
	  Mlast =    array(0,dimW)
	  Qlast =    array(0,dimW)
	  Mcurrent = array(0,dimW)
	  Qcurrent = array(0,dimW)

	  # adjust y
	  ycorr = y - as.vector(W%*%theta)   # adjust obs by everything before starting iterations
	  # so this is the residual.

	  # use this to set up the starting value for the residual variance
	  vare = crossprod(ycorr)/(nrecords-dimX)
  
	  starttime<-proc.time()
	  timeSNP = 0
	  timetotal=0

	  # mcmc sampling
	  for (iter in 1:numiter) 
	  {  #"START" OF MCMC SAMPLING

	     itersave = iter-burnIn
	     if (itersave ==1) 
	     {
	        alphametrovar = array(0,nSNP)  # reinitialize to 0 post burn-in
	     }

	     #####################################################################################
	     # sample intercept
	     #####################################################################################
	     ycorr    = ycorr + W[,1]*theta[1]
	     rhs      = sum(ycorr)/vare        # 1'(y-W*theta)/sigmae
	     invLhs   = 1.0/(nrecords/vare)    # 1/(sigmae/n)		
	     meancond = rhs*invLhs                          
	     theta[1] = rnorm(1,meancond,sqrt(invLhs))       
	     ycorr    = ycorr - W[,1]*theta[1]  # take it off again to create new "residual"	

	     #####################################################################################	
	     # sample variance & effect for each SNP	
	     ####################################################################################
	     if (pi == 1)  
	     { # "START" OF BAYESA
	       #####################################################################################
	       # sample variance for each SNP (Bayes A)	
	       #####################################################################################
	       varq = (scalea*def + SNPeff*SNPeff)/rchisq(nSNP,def+1)  # joint sample 
	       #####################################################################################
	       # sample effect for each SNP (Bayes A)	
	       #####################################################################################
	       startSNP = proc.time()
	       BayesC<- .Call("BayesAC",nSNP,dimX,nanim,Wstacked,WWdiag, theta, ycorr,varq,vare,SNPeff)
	       theta=BayesC[[1]]
	       ycorr=BayesC[[2]]
	       varq=BayesC[[3]]
	       vare=BayesC[[4]]
	       SNPeff=BayesC[[5]]
	     }   # "END" FOR BAYESA

	     if (itersave>1) 
	     {     # "START" OF COMPUTE POSTERIOR MEANS & VARIANCES
	           Mcurrent = Mlast + (theta-Mlast)/itersave
	           Qcurrent = Qlast + (itersave-1)*((theta-Mlast)^2)/itersave
	           Mlast = Mcurrent
	           Qlast = Qcurrent 
    
	           SNPvar=SNPvarlast + (varq-SNPvarlast)/itersave
	           SNPvarlast=SNPvar
	     }  # "END" OF COMPUTE POSTERIOR MEANS & VARIANCES
	     if (itersave==1) {Mlast = theta;SNPvarlast=varq;} 

	     #####################################################################################	
	     #   sample residual variance
	     #####################################################################################
	     vare = ( crossprod(ycorr) +scale_e )/rchisq(1,nrecords+nu_e)   # prior on vare used
	     #  scale parameter should be e`e + Se,
	     #  degrees of freedom should be n + ve

	     #####################################################################################
	     # sample degrees of freedom
	     #####################################################################################
	     if (!truedef)  
	     {  #"START" OF SAMPLE THE DEGREES OF FREEDOM
	        for (cycle in 1:10){
	        def0=def
	        logdef0 = log(def0)
	        logdef1 = rtnorm(1,logdef0,cdef,upper=log(100))   # proposal log df 
	        def1 = exp(logdef1)
	        lfullcd0 = nSNP*(0.5*def0*(log(scalea)+logdef0-log(2))-lgamma(0.5*def0))
	        lfullcd1 = nSNP*(0.5*def1*(log(scalea)+logdef1-log(2))-lgamma(0.5*def1))
	        lfullcd0 = lfullcd0 - (0.5*def0+1)*sum(log(varq[which(varq>0)]))-(0.5*def0*scalea)*sum(1/varq[which(varq>0)]);
	        lfullcd1 = lfullcd1 - (0.5*def1+1)*sum(log(varq[which(varq>0)]))-(0.5*def1*scalea)*sum(1/varq[which(varq>0)]);
	        lfullcd0 = lfullcd0- 2*log(1+def0)+logdef0   # prior + jacobian                                      
	        lfullcd1 = lfullcd1- 2*log(1+def1)+logdef1   # prior + jacobian                          
	        alpha1 = exp(lfullcd1-lfullcd0)
	        alpha1 = min(alpha1,1)
	        if (alpha1 < 1)
	        { 
		  u = runif(1)
		  if (u<alpha1) {def = def1}
		  else def = def0
	        }
	       else {def = def1;}
	       alphatot1=alphatot1+alpha1
	       if ((cycle%%skip == 0) & (iter <= burnIn))
	       {
	         alphacheck = alphatot1/10;
	         if (alphacheck > .70) { cdef = cdef*1.2}
	         else if(alphacheck  < .20) {cdef = cdef*.7}     # tune Metropolis sampler #
	         alphatot1 = 0;
	       }
	       else if ((cycle%%skip == 0) & (iter > burnIn)) 
	       {
	         alphacheck = alphatot1/skip;
	         alphadef_save[itersave/skip] = alphacheck
	         alphatot1 = 0
	       }
	       }
	     }   # "END" OF SAMPLE THE DEGREES OF FREEDOM  

	     #####################################################################################
	     #  sample scale#
	     #  based on Gelman Uniform(0,A) prior on scale parameter for "large" A.
	     #####################################################################################
	     if (!truescale) 
	      {  #"START" OF SAMPLE THE SCALE PARAMETER 
	           shape_g = 0.5*(nSNP*def+1)
	           scale_g = 0.5*def*sum(1/varq)
	           scalea = rgamma(1,shape=shape_g,rate=scale_g)
	      } #"END" OF SAMPLE THE SCALE PARAMETER 

	       ###########################################################################################
	       #store MCMC samples at every "skip"
	       ###########################################################################################
	       if (iter%%skip == 0) 
	        {
	            varesave[iter/skip] = vare
	            if(!truedef) {defsave[iter/skip] = def}
	            if(!truescale) {scalesave[iter/skip] = scalea}
	            if(!truepi) {pisave[iter/skip] = pi}
	        }

	       endSNP = proc.time()
	       timeSNP = timeSNP + (endSNP[1]-startSNP[1]) 

	       endtime = proc.time()
	       timetotal = timetotal + (endtime[1]-starttime[1])

	       if (iter%%1000==0) 
	       {

	       #############################################################################################
	       #    PROCESSING MCMC SAMPLES
	       #############################################################################################
	       hyperparameters = as.matrix(varesave)
	       colnames(hyperparameters) = "varesave"
	       if (!truescale) {hyperparameters = cbind(hyperparameters,scalesave) }
	       if (!truedef) {hyperparameters = cbind(hyperparameters,defsave) }
	       if (!truepi) {hyperparameters = cbind(hyperparameters,pisave) }
	       rownames(hyperparameters) = seq(skip,numiter,skip)
	       save(iter,hyperparameters,file=paste(Save.at, "Hyperparameters",Seed,".RData",sep="")) 
 
	       SNPtimepercycle = timeSNP/iter
	       timepercycle=timetotal/iter
	       save(iter,timepercycle,SNPtimepercycle,file=paste(Save.at, "time",Seed,".RData",sep=""))

	       }# "END" OF PROCESSING MCMC SAMPLING
               
	       if (iter%%1000==0 & itersave>1 ) 
	       {

	       #############################################################################################
	       #    PROCESSING MCMC SAMPLES
	       #############################################################################################
	       meanmu=Mcurrent[dimX]
	       meang= Mcurrent[(dimX+1):dimW]
	       varg = Qcurrent[(dimX+1):dimW]/itersave
	       save(iter,meanmu,meang,varg,SNPvar,file = paste(Save.at,"EffectsResults",Seed,".RData",sep=""))

	       #  monitor MH acceptance rates over cycle for the degrees of freedom term.
	       save(iter,alphadef_save,file=paste(Save.at, "alphadef",Seed,".RData",sep=""))

	       } # "END" OF PROCESSING MCMC SAMPLING

	  } # "END" OF MCMC SAMPLING

}


lib/iPlant/RIO.pm
package iPlant::RIO;

use 5.006;
use strict;
use warnings FATAL => 'all';
use JSON;

=head1 NAME

iPlant::RIO - The great new iPlant::RIO!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Factory to generate R code blocks to handle input/output functions

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.


=head1 SUBROUTINES/METHODS

=head2 make_script
Generates a self contained R script
=cut


my$TABLE_DEFAULT=>{
	'header'=>'TRUE',
	'sep'=>'/t',
	'quote'=>'',
	'dec'=>'.',
	'row.names'=>'NULL',
	'na.string'=>'NA',
	'fill'=>'TRUE'
};


sub load_spec(){
	my$spec=shift;
	my@ins;
	my@outs;
	my$paras=from_json($spec)->{'parameters'};
	for my$para (@{$paras}){
		if(!$para->{'type'}){
			warn "Parameter ".$para->{'para'}." has no type definition\n";
			next;	
		}
		if($para->{'type'} eq 'input'){	
			my$input=build_input($para);
			push @ins,$input;
		}elsif($para->{'type'} eq 'output'){
			my$output=build_output($para);	
			push @outs,$output;
		}else{
			next;
		}	
	}
	return { 'inputs'=>\@ins,
			'outputs'=>\@outs,
	};
	
}

sub build_input(){
	my$para=shift;
	if(!$para->{'format'}){
		warn "Input parameter ".$para->{'para'}." has no format definition\n";
		next;	
	}
}

sub set_table() {
	
	
	
}


lib/iPlant/RFactory.pm
package iPlant::RFactory;

use 5.006;
use strict;
use warnings FATAL => 'all';
use File::Temp qw/tempdir/;
use Archive::Tar;
use Text::Balanced qw/extract_bracketed/;
use JSON;

#TODO: need to handle multi value params.

=head1 NAME

iPlant::RFactory - The great new iPlant::RFactory!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Factory to generate R code blocks to expose package functions

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.


=head1 SUBROUTINES/METHODS

=head2 make_script
Generates a self contained R script
=cut

sub make_script {
	my $json_string = shift;
	my $json_R      = from_json($json_string);

	my $shebang =
'#!/bin/env Rscript --vanilla --default-packages="datasets, utils, grDevices, graphics, stats,methods"';

	my $note =
'#This script has been generated automagically using iPlant Collaborative RFactory service version '
	  . $VERSION . '
#from '
	  . $json_R->{'origin'}
	  . ' written by '
	  . $json_R->{'author'}
	  . '. Please contact the original author for any questions.
#This script is released under the same copyright terms as '
	  . $json_R->{'origin'} . '
#More information are available at http://github.com/iPlantCollaborativeOpenSource/RFactory';

	my $libs = join ",", @{ $json_R->{'dependencies'} };
	$libs = "\nlibrary(getopt,$libs)\n";

	my $options = _make_opt($json_R);

	my $function;
	if ( !$json_R->{'package'} ) {
		$function =
		    $json_R->{'name'}
		  . '<-function'
		  . $json_R->{'call'}
		  . $json_R->{'fun'};
	}
	else {
		$function = 'library("' . $json_R->{'package'} . '")';
	}

 #	if(!$json_R->{'format'} || lc($json_R->{'format'}) eq 'script' ){
 #		$function=$json_R->{'name'}.'<-function'.$json_R->{'call'}.$json_R->{'fun'};
 #	} elsif (lc($json_R->{'format'}) eq 'file'){
 #		$function='source("'.$json_R->{'origin'}.'")';
 #	} elsif (lc($json_R->{'format'}) eq 'file'){
 #		$function='library("'.$json_R->{'origin'}.'")';
 #	} else{
 #		$json_R->{'name'}.'<-function'.$json_R->{'call'}.$json_R->{'fun'};
 #	}

	my $call = $json_R->{'name'} . $json_R->{'call'};

	my $script = join "\n",
	  ( $shebang, $note, $libs, $options, $function, $call );

	print "$script\n";

}

=head2 _make_opt
Creates the R Getopt and help block based on a parameter json
=cut

sub _make_opt {
	my $json_R = shift;
	my @spec;
	my @defaults;
	my $block;

	my %TYPEOF = (
		'integer'   => 'integer',
		'double'    => 'double',
		'numeric'   => 'double',
		'string'    => 'character',
		'character' => 'character',
		'boolean'   => 'logical',
		'logical'   => 'logical',
		'input'     => 'input'
	);
	print Dumper($json_R);
	#TODO:should eval
	my @params = @{ _make_flags( $json_R->{'parameters'} ) };
	if ( !grep /help/i, @params ) {
		my $help = {
			'para'        => 'help',
			'flag'        => 'h',
			'type'        => 'logical',
			'description' => "Print this help message"
		};
		unshift @params, $help;
	}
	my @options;
	my @descriptions;
	for (@params) {

		my $para     = $_->{'para'};
		my $flag     = $_->{'flag'};
		my $type     = 'character';
		my $required = $_->{'value'} ? 2 : 1;

		if ( $_->{'type'} && $TYPEOF{ lc( $_->{'type'} ) } ) {
			$type = $TYPEOF{ lc( $_->{'type'} ) };
			if ( $type eq 'logical' ) {
				$required = 0;
			}
		}

		if ($required) {
			my $arg;
			if ( $_->{'arg'} ) {

				$arg = $_->{'arg'};
				$arg =~ s/\s+/_/g;
			}
			else {
				$arg = "<$type>";
			}
			if ( $required == 2 ) {
				push @options, "[--$para $arg]";
			}
			else {
				push @options, "--$para $arg";
			}

		}

		#TODO: model input

		my $desc = "\t--$para, -$flag\t" . $_->{'description'};
		if ( defined( $_->{'value'} ) ) {
			$desc .= "(defaults to " . $_->{'value'} . ")";
		}
		push @descriptions, $desc;

		if ( defined( $_->{'value'} ) ) {
			push @defaults, "
if( is.null(opt\$$para)) {
	$para <- " . $_->{'value'} . "
} else {
	$para <- opt\$$para	
}";

		}
		else {
			push @defaults, "$para <- opt\$$para";
		}

		push @spec,
		  "\'$para\',\'$flag\',$required,\'$type\'";    #TODO: model input

	}
	shift @defaults;                                    #removing help
	my $spec = "spec<-matrix(c(\n";
	$spec .= join "\n", @spec;
	$spec .= "\n),byrow=TRUE,ncol=4)\n\nopt<-getopt(spec)\n";

	my $help_text =
	    "Usage: RScript "
	  . $json_R->{'name'} . ".R "
	  . ( join ' ', @options ) . "\n";
	$help_text .= join "\n", @descriptions;

	my $help_block = '
if( ! is.null(opt$help) ){
	cat(\'' . $help_text . '\')
	q("no")	
}';

	$block = join "\n", ( $spec, $help_block, @defaults );

	return $block;
}

sub _make_flags {
	my $params = shift;
	my @alpha = ( 'a' .. 'g', 'i' .. 'z', 'A' .. 'Z' );
	my %alpha;

	for ( my $i = 0 ; $i < @alpha ; $i++ ) {
		$alpha{ $alpha[$i] } = $i;
	}

	my @duplicates;

	my $i = 0;
	for ( @{$params} ) {

		my $flag = lc( substr( $_->{'para'}, 0, 1 ) );
		if ( !$alpha{$flag} ) {
			if ( $alpha{ uc($flag) } ) {
				delete $alpha{ uc($flag) };
				$_->{'flag'} = uc($flag);
			}
			else {
				push @duplicates, $i;
			}
		}
		else {
			delete $alpha{$flag};
			$_->{'flag'} = $flag;
		}
		$i++;
	}
	my @free = keys %alpha;
	for (@duplicates) {
		my $flag = shift @free;
		$params->[$_]->{'flag'} = $flag;
	}

	return $params;
}

=head2 _load_package

Identify and loads all the files that belong to an R package

=cut

use Exporter;
our @ISA = qw(Exporter);

use Cwd;
use Data::Dumper;

sub _load_package {
	my $package_loc = shift;
	my $format      = 'script';
	my @rfiles;
	if ( !-e $package_loc ) {
		warn "$package_loc: No such file or directory.\n";
	}
	if ( $package_loc =~ m/\.tar(?:\.gz)?$/i ) {

		#unzip
		my $tar = Archive::Tar->new($package_loc);

		my $here = cwd();
		my $dir = tempdir( CLEANUP => 0 );
		chdir $dir;

		my @out = $tar->extract();

		chdir $here;

		opendir my $TDIR, $dir
		  or die "Cannot read temporary directory $dir: $!\n";
		my @arc_content = grep !/^\.+/, readdir $TDIR;
		close $TDIR;
		if ( !@arc_content ) {
			warn
"ERROR: $package_loc does not seem to be a valid R package or file\n";
		}
		elsif ( @arc_content == 1 && -d "$dir/$arc_content[0]/" ) {

			$package_loc = "$dir/$arc_content[0]";
		}
		else {
			$package_loc = $dir;
		}

	}

	if ( -d $package_loc ) {
		opendir my $PDIR, $package_loc
		  or die "Cannot open the R package directory $package_loc: $!\n";
		my @arc_content = grep !/^\.+/, readdir $PDIR;
		close $PDIR;
		my @rdir = grep /^R$/, @arc_content;
		my $rdir = @rdir ? $rdir[0] : '';
		if ( $rdir && -d "$package_loc/$rdir" )
		{    #it's a well structired package

			#TODO:Read info from DESCRIPTION file
			opendir my $RDIR, "$package_loc/$rdir"
			  or die
			  "Cannot open the R package directory $package_loc/$rdir: $!\n";
			@rfiles = grep /\.[RSqrs]$/, readdir $RDIR;
			close $RDIR;
			for ( my $i = 0 ; $i < @rfiles ; $i++ ) {
				$rfiles[$i] = "$package_loc/R/$rfiles[$i]";
			}

		}
		else {
			@rfiles = grep /\.[RSqrs]$/, @arc_content;
			for ( my $i = 0 ; $i < @rfiles ; $i++ ) {
				$rfiles[$i] = "$package_loc/$rfiles[$i]";
			}
		}

	}
	elsif ( $package_loc =~ m/\.[RSqrs]$/ ) {    #it's a single R file
		push @rfiles, $package_loc;
	}

	if ( !@rfiles ) {
		warn
		  "ERROR: $package_loc does not seem to be a valid R package or file\n";
	}
	return \@rfiles;
}

=head2 _parse_R

=cut

sub _parse_R {
	my @functions;

	#load file
	my $in_file = shift;
	my %r_content;
	open my $RFILE, "<$in_file" or die "Cannot load input R file: $!\n";
	my $r_file = join "", grep !/^\s*#.*$/, (<$RFILE>);
	close $RFILE;

	$r_file =~ s/(?:\s*#.*$)//gm;    #|(?:^\s+)
	$r_file =~ s/\s*\n+/\n/g;

	my @libs;

	#identify libraries
	while ( $r_file =~ s/library\(((?:\w,?)+)\)//gx ) {
		push @libs, split /,/, $1;
	}

	#deduplicate
	my %libs;
	for (@libs) {
		$libs{$_} = 1;
	}

	#identify functions
	my $remainder = $r_file;
	while ( $remainder =~ m/(\w+)\s*(?:=|<-)\s*function/xg ) {
		my %fun;
		$fun{'name'} = $1;
		( my $call, $remainder ) = extract_bracketed( $remainder, '(' );
		$fun{'call'} = $call;
		$call =~ s/(?:,?\s*\w*\s*\()/,/xg;
		$call =~ s/\)|\s+//g;

		#		$call=~ s/(\w+)\s*=\s*(\w+)\s*\((.*)\)/
		my @para = grep !/^\s*$/, split /,/, $call;
		$fun{'para'} = _extract_para( \@para );

		( my $fun, $remainder ) = extract_bracketed( $remainder, '{' );

		#				$fun = '{OK...}';
		$fun{'fun'} = $fun;
		push @functions, \%fun;
		$fun{'dependencies'} = [ keys %libs ];
		$fun{'origin'}       = $in_file;
	}
	print to_json( \@functions );
	return \@functions;
}

=head2 _extract_para

=cut

sub _extract_para {

	#extract parameters
	my @para_list = @{ shift() };
	my @para_objs;
	for (@para_list) {
		my %para_obj;
		my ( $para, $def ) = split /=/, $_;
		%para_obj = (
			'para'        => $para,
			'description' => '',
			'type'        => '',
			'arg'         => '',
		);

		if ($def) {
			$para_obj{'value'} = $def;
			if ( $def =~ /^"(\S+)"$/ ) {
				$para_obj{'type'} = 'string';
			}
			elsif ( $def =~ /^(?:(?:TRUE)|(?:FALSE)|T|F)$/ ) {
				$para_obj{'type'} = 'logical';
			}
			elsif ( $def =~ /^(\d+)$/ ) {
				$para_obj{'type'} = 'integer';
			}
			elsif ( $def =~ /^(\d*.\d+)$/ ) {
				$para_obj{'type'} = 'double';
			}
		}

		push @para_objs, \%para_obj;
	}
	return \@para_objs;
}

=head1 AUTHOR

Naim Matasci, C<< <nmatasci at iplantcollaborative.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-iplant-rfactory at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=iPlant-RFactory>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc iPlant::RFactory


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=iPlant-RFactory>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/iPlant-RFactory>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/iPlant-RFactory>

=item * Search CPAN

L<http://search.cpan.org/dist/iPlant-RFactory/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Naim Matasci.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1;    # End of iPlant::RFactory
